### ISSUE TITLE
Websocket compression optimisation for sending duplicate messages to large number of clients

### ISSUE THREAD

**Issue by JayDadhania** [ninenines/cowboy#1659]
UseCase: 
In a public data socket server like weather data or financial market data almost all the clients are subscribed to the same data. Enabling compression in such a server compresses payloads separately for all the client processes. So one message being sent to 1000 clients will do the same process of compressing that one message 1000 times. 

Possible Solution:
The message can be prepared before being broadcasted to all the connection processes and cowboy can handle such messages separately, being sent as compressed message without compressing each one of them separately.

I don't have the expertise to implement such a feature, but I might be able to help if someone can guide me. I'm very ignorant of the websocket protocol details around compression. I tried to build a work around for this in my server by keeping `compress: false` and mimicking compression by sending relevant headers for clients who support compression and sending them precompressed messages. However I was not successful in doing so as I was getting bad frame errors. I had expected this but it was worth a try. 

I think this is a very common usecase for any socket server. Any help here is appreciated.

**Comment by essen** [comment]
I don't think you can easily do this because the compression is at the protocol level and has to be negotiated. This means that clients that don't negotiate it need to receive data uncompressed, and that clients may negotiate different levels of compression depending on their needs.

You may implement what you want by instead compressing at the application level, meaning that you would use binary frames and send compressed data on them and expect the clients to understand that and decompress them. This understanding is signaled by subprotocol negotiation: https://datatracker.ietf.org/doc/html/rfc6455#section-1.9 - in your case you would define an application-specific subprotocol that does what you want.

This approach is similar to how you would compress data that goes through something like RabbitMQ, at the application level, except using subprotocols instead of content-types to indicate the data is compressed.

**Comment by essen** [comment]
I believe this is answered. Closing, thanks!
